File: LeetCode/Python3/LC20.py
"""
LeetCode Problem - 20
Allan Hieng
"""

def isValid(str):

    if len(str) == 0:
        return True

    stack = []
    for char in str:
        if char == '(':
            stack.append(char)
        elif char == '[':
            stack.append(char)
        elif char == '{':
            stack.append(char)
        else:
            if len(stack) == 0:
                return False
            leftBracket = stack.pop()
            if leftBracket == '[':
                if char != ']':
                    return False
            if leftBracket == '(':
                if char!= ')':
                    return False
            if leftBracket == '{':
                if char!= '}':
                    return False
    
    if len(stack) != 0:
        return False
    else:
        return True

File: LeetCode/C++/LC20.cpp
#include <stack>
#include <string>

using namespace std;

class Solution {
    public:
        bool isValid(string s){
            stack <char> para;

            if (s.length() == 0){
                return true;
            };

            for (int i = 0; i < s.length(); i++){
                if (s.at(i) == '('){
                    para.push(s.at(i));
                } else if (s.at(i) == '{'){
                    para.push(s.at(i));
                } else if (s.at(i) == '['){
                    para.push(s.at(i));
                } else {
                    if (para.empty()){
                        return false;
                    }
                    char leftBracket = para.top();
                    para.pop();
                    if (leftBracket == '('){
                        if (s.at(i) != ')'){
                            return false;
                        };
                    } else if (leftBracket == '{'){
                        if (s.at(i) != '}'){
                            return false;
                        };
                    } else if (leftBracket == '['){
                        if (s.at(i) != ']') {
                            return false;
                        };
                    }
                };
            };
            if (para.empty()){
                return true;
            } else {
                return false;
            };
        };
};